### PR TITLE
RPS

### PR DESCRIPTION
**Pull request by harrythomson1**
**Pros**

- Happy with my OOP and SRP. I toyed around with the idea of having another class to display the result but ultimately decided against it.
- Finished both user stories!
- Full functionality
- Replaced all global variables I originally used with sessions once I had read a bit more into how they work
- Really happy with my tests, used both doubles and stubs

**Cons**

- It is UGLY. I didn't style any of it
- Didn't even attempt the stretch goals
- app.rb file is a mess
- I didn't create any .self methods of my classes as I don't understand them but I believe they'd be better and cleaner than using sessions
- Rock paper scissors logic is complex within my Game class
- My tests aren't very DRY I definitely could've used a ruby file with a method in it where I have visited the home page, put in my name and pressed a button as I used this code on nearly all tests
- Didn't upload to GitHub enough again... Although this time I at least did it once before completing the whole project.

Overall I'm really happy with how this project ended up despite how ugly it is. I focused a lot in my OOP which I think I'm slowly starting to get the concept of but please tell me if you think I'm still not fully understanding the concept.

Thank you for reviewing my code!

### CODE REVIEW

**Comment by harrythomson1** [comment]
So I've since updated my RPS challenge to a V0.2

**Pros**

- Still happy with my OOP, removed a class from player which would allow them to choose rock paper or scissors as this decision is now entirely decided by the button press and stored in a session.
- Finished the stretch goal of adding multiplayer functionality
- Tidied up app.rb but I still think it's a bit of a mess
- Removed 3 pages, I previously had a different page depending on what option was taken all the options now go to a results page and the option chosen is stored within a session which is then used on the Game class
- Added styling, it's still ugly as design is not a strong point of mine but there is some styling there!
- DRY'd up my tests by adding in web helpers
- Uploaded to GitHub much more frequently

**Cons**

- Still no rock paper scissors lizard spock, I think I could implement this functionality but I didn't want to spend anymore of my weekend working on it
- Still not happy with the app.rb file
- Used 'get' pages for the multiplayer functionality these should probably use post instead of get
- Still didn't create any '.self' methods

File: app.rb [hunk 1]
```diff
@@ -1,7 +1,57 @@
 require 'sinatra/base'
+require_relative './lib/computer_choice'
+require_relative './lib/game'
+require_relative './lib/player'
+
```
Comment (EdwardAndress) [inline app.rb:5]: I think I've figured out what's going on but I was slightly confused by the lack of post routes for weapon choice.

File: lib/game.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
+class Game
+
+  def initialize(player_1_choice, player_2_choice)
+    @player_1_choice = player_1_choice.downcase.capitalize
+    @player_2_choice = player_2_choice.downcase.capitalize
+  end
+
```
Comment (EdwardAndress) [inline lib/game.rb:7]: The game logic is duplicated here, in the two play-related methods, which means a change to the rules would always have to involve updating two methods and that might lead to some mistakes.

Can you figure out a way to reduce duplication?

File: spec/computer_choice_spec.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
+require 'computer_choice'
+
+describe ComputerChoice do
+
+  let(:computer) { ComputerChoice.new }
+  
+  describe "rock_paper_scissors" do
+    it 'returns rock' do
+      allow(computer).to receive(:rand) { 1 }
```
Comment (EdwardAndress) [inline spec/computer_choice_spec.rb:9]: Nice use of stubs!

File: spec/features/multiplayer_spec.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  scenario 'player 1 scissors player 2 chooses rock' do
+    multiplayer_sign_in
+    click_button 'Scissors'
+    click_button 'Rock'
+    expect(page).to have_content 'Eloise wins! Rock beats Scissors!'
+  end
+
+  scenario 'player 1 scissors player 2 chooses paper' do
+    multiplayer_sign_in
```
Comment (EdwardAndress) [inline spec/features/multiplayer_spec.rb:55]: nice use of helper methods!

File: spec/features/play_game_spec.rb [hunk 5]
```diff
@@ -0,0 +1,79 @@
+require_relative 'web_helpers'
+
+feature 'play game' do
+
+  let(:computer_choice) { instance_double(ComputerChoice) }
+
+  scenario 'player chooses rock computer chooses rock' do
+    allow(ComputerChoice).to receive(:new) { computer_choice }
```
Comment (EdwardAndress) [inline spec/features/play_game_spec.rb:8]: some nice test isolation going on here!

**Comment by EdwardAndress** [comment]
This is a nice solution, Harry!  I particularly like your use of doubles and stubs.  Once again, your self assessment is also insightful and I agree that there are some places where the codebase could be tidied up.  I've left a few specific comments in line against the code.

I think the general theme is one of reducing repetition.  There might be a way to use an instance of Game to determine whether or not more player input is needed and that might allow you to use the same routes in app.rb for both multiplayer and single player games.  Something like this, for example, when choosing weapons...

```
post '/weapons' do
  game.player_one_weapon = params[:weapon]
  if game.pending_weapon_choice
    @current_player = params[:player_2_name]
    render :weapon # or whatever the weapon choice template is called
  else
    redirect_to '/result'
  end
end
```

actually both branches would probably go to a redirect but I wanted to keep the example as brief as possible

**Pull request closed without merging**